### ISSUE TITLE
Chapter 10: handling multiple outputs-- question about multiple loss weights

### ISSUE THREAD

**Issue by amjass12** [ageron/handson-ml2#136]
Hello @ageron !!

First of all, I just want to say thank you for this book.. one thing that is particularly unique is how easy it is to understand very complex concepts in ML, so thank you for making them so clear!

I am currently implementing a mulit-task model for my project. The construction and running of this model is straightforward and easy to follow, i have got this to work without issues, however I was wondering if you could explain in a little more detail, the `loss_weights` when there are multiple losses. One thing that isn't clear to me (and i have looked online and it still isn't clear) is the balance of setting weights for losses when there are multiple outputs. 

I have a multi-task model that takes in one input and has two output nodes. One that predicts many classes and labels (so multi-class multi-label, binary_crossentropy) and one output node where one class is known, and I would like the network to predict ONE of those classes (or highest score wins approach) so, one class --categorical_crossentropy. 

My question is linked to the argument in model.compile:

`
model.compile(loss= ['binary_crossentropy', 'categrocial_crossentropy'], loss_weights= [?,?], optimizer='adam')
`
This is one point of clarification i would really like. I understand that by default, keras assigns an equal weighting for each loss... what i am unsure about is how the balance should be split, and i am unsure because i am unsure about what `loss_weights` actually means.. whether it is for example, an importance value? please could you expand on this?

If it is an importance ratio, I would assume that i should leave this argument as the default and maintain equality for both losses? -- If i understand it correctly based on my reading, it is how much each loss contributes to the total loss value? In this case, again i would assume equality between the two?

presumably, the individual losses of each node will not change whatever the case is?

and hence why in the book, the aux_output has a lower weighting?

thank you so much!

**Comment by ageron** [comment]
Hi @amjass12 ,

Thanks for your question, and sorry for the late response. I think you're very close to understanding everything correctly: the loss weights are indeed a measure of the relative importance of each output. Let me try to clarify this.

Gradient Descent works by tweaking the model parameters to bring the loss down. If your model has multiple heads (in your example, two heads), then you will have one loss function per head (in your case, `binary_crossentropy` and `categorical_crossentropy`), so they must be combined into a single loss function in order to get Gradient Descent to work. If we have _n_ heads, with losses L<sub>1</sub>, L<sub>2</sub>, ..., L<sub>_n_</sub>, then the simplest way to combine them into a single loss is just to sum them: L = L<sub>1</sub> + L<sub>2</sub> + ... + L<sub>_n_</sub>. But this gives every head the same importance (as far as Gradient Descent is concerned). Sometimes, you may believe that one output is more important than others. In that case, you can put more weight onto it. So the loss function becomes: L = &alpha;<sub>1</sub>L<sub>1</sub> + &alpha;<sub>2</sub>L<sub>2</sub> + ... + &alpha;<sub>_n_</sub>L<sub>_n_</sub> where &alpha;<sub>1</sub>, &alpha;<sub>2</sub>, ..., &alpha;<sub>_n_</sub> are the loss weights.

Let's look at an example. Suppose you are implementing a model to classify and locate an animal in a picture. If the model makes an imperfect prediction, would you prefer it to identify the right animal but in the wrong place, or would you prefer it to identify the wrong animal but in the right place? Of course, this depends on the application: you can easily imagine applications where the getting the animal right is most important, and other applications where getting the position right is most important).  This is what the `loss_weight` allows you to tweak. It depends heavily on your business objective.

If one of the outputs is a regularization output, then it's a bit different. For example, suppose you're training an image classification model, with two outputs: the main classification output, and an auxiliary regularization output. The regularization output is only used during training. What you really care about at the end of the day is that the model's accuracy is as high as possible on the main output on new images after training. So you can just treat the `loss_weights` as a hyperparameter, and just pick the values that lead to the highest model accuracy on the validation set (on the main output). For example, suppose you train the model with `loss_weights=[1.0, 1.0]` (which is the default) and after training you get 90% accuracy on the validation set (on the main output), then you train the model again with `loss_weights=[1.0, 0.5]` and you get 95% accuracy on the validation set, then you should prefer the second option (of course you should try many more values and pick the best one).

I hope this helps.

**Comment by amjass12** [comment]
Hi @ageron 

Thank you so much for taking the time to answer my question in detail. Your explanation makes perfect sense and is perfectly clear to me!!

I will treat the weights as a hyper-parameter and explore. 

thanks again!
